### PR TITLE
UI: Add downstream keyer

### PR DESCRIPTION
**Pull request by cg2121**
This adds the ability for a user to select a scene to be used
as a downstream keyer.

Current issues:
-~~the DSK scene is not always on top~~
-~~the DSK scene item shows up in sources list~~
-~~crash on exit~~

### CODE REVIEW

**Comment by SuslikV** [comment]
I can forget to enable/disable hotkeys for it, or I cannot?

**Comment by cg2121** [comment]
I believe I have fixed all of the issues now.

**Comment by SuslikV** [comment]
Blind faith...
I see, auto-building under the mac don't know this Member Function:
https://github.com/obsproject/obs-studio/pull/1321/files#diff-a0eec92bde9402a347f40d7593c0cdc9R3584
or `QAction::QAction(QString)`

File: UI/data/locale/en-US.ini [hunk 1]
```diff
@@ -845,3 +845,8 @@ OutputWarnings.MP4Recording="Warning: Recordings saved to MP4 will be unrecovera
 # deleting final scene
 FinalScene.Title="Delete Scene"
 FinalScene.Text="There needs to be at least one scene."
+
+# Dowstream Keyer
```
Comment (RytoEX) [inline UI/data/locale/en-US.ini]: Spelling:  "Dowstream" should be "Downstream"

File: UI/data/locale/en-US.ini [hunk 2]
```diff
@@ -845,3 +845,8 @@ OutputWarnings.MP4Recording="Warning: Recordings saved to MP4 will be unrecovera
...
+DownstreamKeyerScene="Set as Downstream Keyer"
+DownstreamKeyerReset="Reset Downstream Keyer"
+DownstreamKeyerExclude="Exclude from downstream keyer"
```
Comment (RytoEX) [inline UI/data/locale/en-US.ini]: Everywhere else has "Downstream Keyer" capitalized as such.  Is there a particular reason to lowercase it here?

File: UI/window-basic-main-scene-collections.cpp [hunk 3]
```diff
@@ -244,7 +244,6 @@ void OBSBasic::RefreshSceneCollections()
 	ui->actionRemoveSceneCollection->setEnabled(count > 1);
 
 	OBSBasic *main = reinterpret_cast<OBSBasic*>(App()->GetMainWindow());
-
```
Comment (RytoEX) [inline UI/window-basic-main-scene-collections.cpp]: This whitespace change seems unrelated to the rest of the PR.  You should revert this.

File: UI/window-basic-main.cpp [hunk 4]
```diff
@@ -2551,6 +2614,7 @@ void OBSBasic::ActivateAudioSource(OBSSource source)
 
 	bool vertical = config_get_bool(GetGlobalConfig(), "BasicWindow",
 			"VerticalVolControl");
+
```
Comment (RytoEX) [inline UI/window-basic-main.cpp]: This whitespace change seems unrelated to the rest of the PR. You should revert this.

File: UI/window-basic-main.cpp [hunk 5]
```diff
@@ -4145,6 +4224,12 @@ void OBSBasic::on_actionSourceProperties_triggered()
 
 void OBSBasic::on_actionSourceUp_triggered()
 {
+	QModelIndexList selectedItems =
+		ui->sources->selectionModel()->selectedIndexes();
```
Comment (RytoEX) [inline UI/window-basic-main.cpp]: Style:  Use two tabs to indent from the original line.

**Comment by jp9000** [comment]
I do want to see DSK implemented, though I want to flesh out some better design ideas for it first.  I think we can possibly make this a bit better with more features.

It is however a very clever way of solving the problem with the existing system.  The only problem is that no one ever right-clicks on scenes, so almost no one would probably ever realize that the feature exists -- that sort of UX issue is another issue in and of itself.

**Comment by cg2121** [comment]
I now pushed the refactor.

TODO:
-Fix crash on exit
-Figure out why the dsk is not saving and loading

**Comment by cg2121** [comment]
Just need to fix a few bugs.

**Comment by cg2121** [comment]
I'm closing this until I can figure out newly discovered issues with saving and loading.

File: libobs/obs-source.c [hunk 6]
```diff
@@ -366,16 +368,17 @@ static obs_source_t *obs_source_create_internal(const char *id,
 }
 
 obs_source_t *obs_source_create(const char *id, const char *name,
-		obs_data_t *settings, obs_data_t *hotkey_data)
+		obs_data_t *settings, obs_data_t *hotkey_data, bool dsk)
```
Comment (WizardCM) [inline libobs/obs-source.c]: Note: this seems to break obs-ndi. Would this be considered a breaking change to the API?

"'obs_source_create': function does not take 4 arguments"

Comment (RytoEX) [reply libobs/obs-source.c]: If it changes the API and external things (plugins) rely on that API, and the change causes those plugins to break, then yes, I'd say that's a breaking change.  In semver, such a change would be reserved for a major release like OBS 24.0.0 (or libobs 24.0.0, which is tied to OBS releases).

Comment (WizardCM) [reply libobs/obs-source.c]: Then I would suggest a backwards compatible helper function that accepts 4 arguments and passes false for the last parameter on its own.

**Comment by cg2121** [comment]
I have now rebased to master and applied clang formatting.

**Comment by WizardCM** [comment]
Friendly bump: there are merge conflicts. Though I do believe more of a discussion is needed in terms of UX and UI, correct?

**Comment by mufunyo** [comment]
I mentioned this on Discord months and months ago, but I don't want it lost in the ether so I'll repost it here in the PR for posterity:

I'd suggest changing the UI from a traditional DSK workflow (which would be more at home in something like Showrunner), to something more generic that fits the scenes/sources paradigm of OBS better.

By pinning a source (shown in the sources list with a pushpin📌 icon) you can make it show up in every scene. By pinning a scene you can make it show up in every scene other than itself or other pinned scenes (to prevent recursion).

Once a source or scene has been pinned, its visibility can be toggled, which also toggles it for all other scenes. Unpinning a source removes it from every scene other than the scene currently being edited. Pinning or unpinning a source or scene does not make it appear/disappear in the program feed instantly when using studio mode; you have to transition from preview to program to take the new pin state live.

**Comment by jp9000** [comment]
Having DSKs would be awesome, but I disagree with the approach taken.  I
don't think you need to modify libobs in order to even implement DSKs,
as libobs already has the ability to display multiple channels on top of
one another at a time.  It's still not going to be simple, but I really
don't think it necessitates modifying how sources are created and such.

If you have time, write an RFC for this and we'll hammer out the details
before we start attempting to work on something like this.  It really
needs further discussion.

**Comment by summershrimp** [comment]
Any future update for this? It's really necessary to have DSK feature for production live.
